src/wallet/wallet.cpp

```diff
@@ -82,13 +82,46 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
     return nullptr;
 }
 
+static Mutex g_wallet_release_mutex;
+static std::condition_variable g_wallet_release_cv;
+static std::set<CWallet*> g_unloading_wallet_set;
+
 // Custom deleter for shared_ptr<CWallet>.
 static void ReleaseWallet(CWallet* wallet)
 {
-    wallet->WalletLogPrintf("Releasing wallet\n");
-    wallet->BlockUntilSyncedToCurrentChain();
-    wallet->Flush();
-    delete wallet;
+    {
+        std::unique_lock<std::mutex> lock(g_wallet_release_mutex);
+        // Assert that there was an explict UnloadWallet call.
+        assert(g_unloading_wallet_set.erase(wallet) == 1);
```

Comment:
Would suggest:

```c++
size_t erased = g_unloading_wallet_set.erase(wallet) == 1
assert(erased == 1);
```

Even though we don't support building bitcoin with NDEBUG, I think most people just expect asserts to contain error checking, not code that is important to run.